src/main/java/ai/ancf/lmos/operator/reconciler/AgentReconciler.java

```diff
@@ -6,28 +6,66 @@
 
 package ai.ancf.lmos.operator.reconciler;
 
-import io.javaoperatorsdk.operator.api.reconciler.Context;
-import io.javaoperatorsdk.operator.api.reconciler.ControllerConfiguration;
-import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
-import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
-import ai.ancf.lmos.operator.resources.agent.AgentResource;
+import ai.ancf.lmos.operator.service.AgentReconciliationService;
+import ai.ancf.lmos.operator.service.AgentServiceQuery;
+import ai.ancf.lmos.operator.service.KubernetesResourceManager;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.javaoperatorsdk.operator.api.reconciler.*;
+import org.eclipse.ditto.json.JsonObject;
+import org.eclipse.ditto.wot.model.ThingDescription;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
 
+import java.util.concurrent.TimeUnit;
 
+/**
+ * Reconciles Deployment resources by watching associated Pods and registering Services.
+ */
 @Component
-@ControllerConfiguration
-public class AgentReconciler implements Reconciler<AgentResource> {
+@ControllerConfiguration(labelSelector = "wot-agent=true")
+public class AgentReconciler implements Reconciler<Deployment>, Cleaner<Deployment> {
 
     private static final Logger LOG = LoggerFactory.getLogger(AgentReconciler.class);
 
+    private final AgentServiceQuery agentServiceQuery;
+    private final AgentReconciliationService agentReconciliationService;
+    private final KubernetesResourceManager kubernetesResourceManager;
+
+    public AgentReconciler(AgentServiceQuery agentServiceQuery, AgentReconciliationService agentReconciliationService, KubernetesResourceManager kubernetesResourceManager) {
+        this.agentServiceQuery = agentServiceQuery;
+        this.agentReconciliationService = agentReconciliationService;
+        this.kubernetesResourceManager = kubernetesResourceManager;
+    }
 
     @Override
-    public UpdateControl<AgentResource> reconcile(AgentResource agentResource, Context context) {
-        // TODO: fill in logic
-        LOG.debug("Agent reconcile");
+    public UpdateControl<Deployment> reconcile(Deployment deployment, Context context) {
+
+        boolean isDeploymentReady = agentReconciliationService.isDeploymentReady(deployment);
+
+        LOG.info("is Deployment {} ready: {}", deployment.getMetadata().getName(), isDeploymentReady);
 
-        return UpdateControl.noUpdate();
+        if(isDeploymentReady) {
+            try {
+                String serviceUrl = kubernetesResourceManager.getServiceUrl(deployment);
+                String tdString = agentServiceQuery.queryAgentService(serviceUrl);
+                            ThingDescription thingDescription = ThingDescription.fromJson(JsonObject.of(tdString));
+                            assert(thingDescription.getDescription().isPresent());
```

Comment:
I think we can remove this here

Reply:
assertion removed, validation moved to separate method. Description is still mandatory in agent response

Reply:
resolved